File: EPIJudge/epi_judge_cpp/lowest_common_ancestor.cc
#include <memory>

#include "binary_tree_node.h"
#include "test_framework/binary_tree_utils.h"
#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/timed_executor.h"
using std::unique_ptr;

struct State {
    int matches = 0;
    BinaryTreeNode<int> *node;
};

State LCA(const unique_ptr<BinaryTreeNode<int>>& tree,
          const unique_ptr<BinaryTreeNode<int>>& node0,
          const unique_ptr<BinaryTreeNode<int>>& node1) {
    State next = {0, nullptr};
    if (tree == nullptr) {
        return next;
    }
    State left = LCA(tree->left, node0, node1);
    if (left.matches == 2) return left;
    State right = LCA(tree->right, node0, node1);
    if (right.matches == 2) return right;
    next.matches = left.matches + right.matches + (tree == node0) + (tree == node1);
    if (next.matches == 2) next.node = tree.get();
    return next;
}

BinaryTreeNode<int>* Lca(const unique_ptr<BinaryTreeNode<int>>& tree,
                         const unique_ptr<BinaryTreeNode<int>>& node0,
                         const unique_ptr<BinaryTreeNode<int>>& node1) {
    return LCA(tree, node0, node1).node;
}
int LcaWrapper(TimedExecutor& executor,
               const unique_ptr<BinaryTreeNode<int>>& tree, int key0,
               int key1) {
  const unique_ptr<BinaryTreeNode<int>>& node0 = MustFindNode(tree, key0);
  const unique_ptr<BinaryTreeNode<int>>& node1 = MustFindNode(tree, key1);

  auto result = executor.Run([&] { return Lca(tree, node0, node1); });

  if (!result) {
    throw TestFailure("Result can not be nullptr");
  }
  return result->data;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"executor", "tree", "key0", "key1"};
  return GenericTestMain(args, "lowest_common_ancestor.cc",
                         "lowest_common_ancestor.tsv", &LcaWrapper,
                         DefaultComparator{}, param_names);
}

File: EPIJudge/epi_judge_python/lowest_common_ancestor.py
import functools
import collections
from typing import Optional

from binary_tree_node import BinaryTreeNode
from test_framework import generic_test
from test_framework.binary_tree_utils import must_find_node, strip_parent_link
from test_framework.test_failure import TestFailure
from test_framework.test_utils import enable_executor_hook


def lca(tree: BinaryTreeNode, node0: BinaryTreeNode,
        node1: BinaryTreeNode) -> Optional[BinaryTreeNode]:

    State = collections.namedtuple('State', ('n', 'node'))

    def traverse(tree, node0, node1):
        if not tree:
            return State(0, None)

        left = traverse(tree.left, node0, node1)
        if left.n == 2:
            return left

        right = traverse(tree.right, node0, node1)
        if right.n == 2:
            return right

        total = left.n + right.n + int(tree == node0) + int(tree == node1)
        node = None
        if total == 2:
            node = tree
        return State(total, node)

    # TODO - you fill in here.
    return traverse(tree, node0, node1).node


@enable_executor_hook
def lca_wrapper(executor, tree, key1, key2):
    strip_parent_link(tree)
    result = executor.run(
        functools.partial(lca, tree, must_find_node(tree, key1),
                          must_find_node(tree, key2)))

    if result is None:
        raise TestFailure('Result can\'t be None')
    return result.data


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('lowest_common_ancestor.py',
                                       'lowest_common_ancestor.tsv',
                                       lca_wrapper))
